include/nbl/builtin/hlsl/type_traits.hlsl

```diff
@@ -425,4 +400,112 @@ struct is_matrix<matrix<T, N, M> > : bool_constant<true> {};
 }
 }
 }
+
+
+#ifdef __HLSL_VERSION
+
+
+// DXC doesn't support linking SPIR-V so this will always work I guess?
+// split because we won't be able to use `typeid` or `decltype` on functions until https://github.com/microsoft/hlsl-specs/issues/100
+#define NBL_REGISTER_TYPEID(T) namespace nbl { namespace  hlsl { namespace type_traits { namespace impl { \
+    template<> struct typeid_t<T> : integral_constant<uint32_t,__COUNTER__> {}; }}}}
+
+#define NBL_REGISTER_OBJ_TYPE(T) NBL_REGISTER_TYPEID(T); \
+namespace nbl { namespace  hlsl { namespace type_traits {  namespace impl { \
+encoder<typeid_t<T>::value> encode_typeid(T); \
+template<> struct decltype_t<sizeof(encoder<typeid_t<T>::value>)/4> \
+{ using type = T; }; }}}}
+
+#define typeid(expr) (sizeof(::nbl::hlsl::type_traits::impl::encode_typeid(expr))/4)
+#define decltype(expr) ::nbl::hlsl::type_traits::impl::decltype_t<typeid(expr)>::type
+
+// builtins
+NBL_REGISTER_OBJ_TYPE(int16_t)
+NBL_REGISTER_OBJ_TYPE(int64_t)
+
+NBL_REGISTER_OBJ_TYPE(uint16_t)
+NBL_REGISTER_OBJ_TYPE(uint64_t)
+
+NBL_REGISTER_OBJ_TYPE(bool)
+NBL_REGISTER_OBJ_TYPE(bool4)
+NBL_REGISTER_OBJ_TYPE(bool3)
+NBL_REGISTER_OBJ_TYPE(bool2)
+NBL_REGISTER_OBJ_TYPE(bool1)
+
+NBL_REGISTER_OBJ_TYPE(int32_t)
+NBL_REGISTER_OBJ_TYPE(int32_t4)
+NBL_REGISTER_OBJ_TYPE(int32_t3)
+NBL_REGISTER_OBJ_TYPE(int32_t2)
+NBL_REGISTER_OBJ_TYPE(int32_t1)
+
+NBL_REGISTER_OBJ_TYPE(uint32_t)
+NBL_REGISTER_OBJ_TYPE(uint32_t4)
+NBL_REGISTER_OBJ_TYPE(uint32_t3)
+NBL_REGISTER_OBJ_TYPE(uint32_t2)
+NBL_REGISTER_OBJ_TYPE(uint32_t1)
```

Comment:
I think these also come in 16 and 64 bit flavours....

maybe you're like to make a `#define NBL_REGISTER_VECTORS_AND_MATRICES(T)` with `BOOST_PP_FOR_EACH` ?
(likeweise for the C++ compat crap)

Reply:
Hmm using boost in hlsl side requires us to append the headers into NBL_RESOURCES_TO_EMBED in cmakelists, will look into that later

Reply:
> Hmm using boost in hlsl side requires us to append the headers into NBL_RESOURCES_TO_EMBED in cmakelists, will look into that later

boost is already there, [embedded as separate library](https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/3rdparty/boost/CMakeLists.txt#L17) linked within Nabla and [mounted into our filesystem](https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/src/nbl/system/ISystem.cpp#L25), [here](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/blob/3d14f7f50c980094715734f3da80c4792d692ef7/52.SystemTest/main.cpp#L308) is an example

Reply:
> Hmm using boost in hlsl side requires us to append the headers into NBL_RESOURCES_TO_EMBED in cmakelists, will look into that later

they're already there AFAIK, @AnastaZIuk did it

Reply:
but boost PP not needed , solution you made is good

Reply:
HLSL compiler fails when I try to include in shaders though
```
[07.10.2023 16:08:53:946922][ERROR]: DXC Compilation Failed:
boost/preprocessor.hpp:17:11: fatal error: 'boost/preprocessor/library.hpp' file not found
# include <boost/preprocessor/library.hpp>
```